### ISSUE TITLE
flatpak enter: error: Can't enter user namespace: Invalid argument

### ISSUE THREAD

**Issue by devurandom** [flatpak/flatpak#3406]
## Linux distribution and version

Gentoo Linux

## Flatpak version

```
❯ flatpak --version
Flatpak 1.6.1
```

## Description of the problem

```
❯ flatpak ps
Instance   PID    Application             Runtime
2160942109 357295 com.valvesoftware.Steam org.freedesktop.Platform
[...]

❯ flatpak enter -vv 2160942109 /bin/sh
error: Can't enter user namespace: Invalid argument

❯ flatpak enter -vv 357295 /bin/sh
error: Can't enter user namespace: Invalid argument

❯ sudo -E flatpak enter -vv 2160942109 /bin/sh
error: Can't enter user namespace: Invalid argument

❯ sudo -E flatpak enter -vv 357295 /bin/sh
error: Can't enter user namespace: Invalid argument

❯ sudo flatpak enter -vv 2160942109 /bin/sh
error: 2160942109 is neither a pid nor an application or instance ID

❯ sudo flatpak enter -vv 357295 /bin/sh
[📦 com.valvesoftware.Steam ~]# test -d /app && echo in container || echo broken
broken
```

## Steps to reproduce

Not really sure. I tried it only on my own system.

## Extra information

```
Portage 2.3.87 (python 3.6.9-final-0, default/linux/amd64/17.1/desktop/plasma/systemd, gcc-9.2.0, glibc-2.30-r3, 5.4.17 x86_64)
=================================================================
                         System Settings
=================================================================
System uname: Linux-5.4.17-x86_64-AMD_Ryzen_5_2400G_with_Radeon_Vega_Graphics-with-gentoo-2.6
KiB Mem:    14134704 total,   1572828 free
KiB Swap:          0 total,         0 free
Timestamp of repository gentoo: Tue, 04 Feb 2020 17:45:01 +0000
Head commit of repository gentoo: 29eb6d9689cb18494c32a24bf03b25bfc1168b65
Head commit of repository flatpak-overlay: 5a60266f5803ac62b58994da0020d101fe424dd7

Head commit of repository local: 178e4c57d196a3f1573a5ee583c88ecfe8079842

sh bash 5.0_p11
ld GNU gold (Gentoo 2.33.1 p2 2.33.1) 1.16
ccache version 3.7.7 [disabled]
app-shells/bash:          5.0_p11::gentoo
dev-java/java-config:     2.2.0-r4::gentoo
dev-lang/perl:            5.30.1::gentoo
dev-lang/python:          2.7.17-r1::gentoo, 3.6.10::gentoo, 3.7.6::gentoo, 3.8.1::gentoo, 3.9.0_alpha2::gentoo
dev-util/ccache:          3.7.7::gentoo
dev-util/cmake:           3.16.3::gentoo
dev-util/pkgconfig:       0.29.2::gentoo
sys-apps/baselayout:      2.6-r1::gentoo
sys-apps/sandbox:         2.18::gentoo
sys-devel/autoconf:       2.13-r1::gentoo, 2.69-r5::gentoo
sys-devel/automake:       1.13.4-r2::gentoo, 1.16.1-r2::gentoo
sys-devel/binutils:       2.33.1-r1::gentoo, 2.34::gentoo
sys-devel/gcc:            9.2.0-r3::gentoo
sys-devel/gcc-config:     2.2.1::gentoo
sys-devel/libtool:        2.4.6-r6::gentoo
sys-devel/make:           4.3::gentoo
sys-kernel/linux-headers: 5.5::gentoo (virtual/os-headers)
sys-libs/glibc:           2.30-r3::gentoo
Repositories:

gentoo
    location: /var/cache/portage/gentoo
    sync-type: rsync
    sync-uri: rsync://rsync.de.gentoo.org/gentoo-portage
    priority: -1000
    sync-rsync-extra-opts: 
    sync-rsync-verify-jobs: 1
    sync-rsync-verify-max-age: 24
    sync-rsync-verify-metamanifest: yes

flatpak-overlay
    location: /var/db/repos/flatpak-overlay
    sync-type: git
    sync-uri: https://github.com/fosero/flatpak-overlay.git
    masters: gentoo

local
    location: /var/cache/portage/local
    sync-type: git
    sync-uri: https://github.com/devurandom/gentoo-overlay.git
    masters: gentoo
    priority: 1000

ACCEPT_KEYWORDS="amd64 ~amd64"
ACCEPT_LICENSE="@FREE"
CBUILD="x86_64-pc-linux-gnu"
CFLAGS="-pipe -O2 -march=znver1"
CHOST="x86_64-pc-linux-gnu"
CONFIG_PROTECT="/etc /etc/grs/systems.conf /usr/lib64/libreoffice/program/sofficerc /usr/share/config /usr/share/gnupg/qualified.txt /usr/share/maven-bin-3.6/conf"
CONFIG_PROTECT_MASK="/etc/ca-certificates.conf /etc/dconf /etc/env.d /etc/fonts/fonts.conf /etc/gconf /etc/gentoo-release /etc/php/apache2-php7.4/ext-active/ /etc/php/cgi-php7.4/ext-active/ /etc/php/cli-php7.4/ext-active/ /etc/revdep-rebuild /etc/sandbox.d /etc/terminfo /etc/texmf/language.dat.d /etc/texmf/language.def.d /etc/texmf/updmap.d /etc/texmf/web2c"
CXXFLAGS="-pipe -O2 -march=znver1"
DISTDIR="/var/cache/portage/distfiles"
EMERGE_DEFAULT_OPTS="--keep-going --nospinner --verbose-conflicts"
ENV_UNSET="DBUS_SESSION_BUS_ADDRESS DISPLAY GOBIN PERL5LIB PERL5OPT PERLPREFIX PERL_CORE PERL_MB_OPT PERL_MM_OPT XAUTHORITY XDG_CACHE_HOME XDG_CONFIG_HOME XDG_DATA_HOME XDG_RUNTIME_DIR"
FCFLAGS="-O2 -pipe"
FEATURES="assume-digests binpkg-docompress binpkg-dostrip binpkg-logs buildsyspkg cgroup compressdebug config-protect-if-modified distlocks ebuild-locks fakeroot fixlafiles ipc-sandbox merge-sync mount-sandbox multilib-strict network-sandbox news parallel-fetch parallel-install pid-sandbox preserve-libs protect-owned sandbox sfperms strict unknown-features-warn unmerge-logs unmerge-orphans userfetch userpriv usersandbox usersync xattr"
FFLAGS="-O2 -pipe"
GENTOO_MIRRORS="http://ftp.spline.inf.fu-berlin.de/mirrors/gentoo/ http://ftp-stud.hs-esslingen.de/pub/Mirrors/gentoo/ http://distfiles.gentoo.org"
LANG="en_US.UTF-8"
LDFLAGS="-Wl,-O1 -Wl,--as-needed -Wl,--hash-style=gnu"
MAKEOPTS="-j6 -l4"
PKGDIR="/var/cache/portage/packages"
PORTAGE_CONFIGROOT="/"
PORTAGE_RSYNC_OPTS="--recursive --links --safe-links --perms --times --omit-dir-times --compress --force --whole-file --delete --stats --human-readable --timeout=180 --exclude=/distfiles --exclude=/local --exclude=/packages --exclude=/.git"
PORTAGE_TMPDIR="/tmp"
USE="7z 7zip X a52 aac aacplus aacs acl acpi activities aio alsa amd64 appindicator appstream archive audit avahi ayatana bash-completion bdplus berkdb blake2 bluetooth bluray branding brotli bs2b btrfs bzip2 cairo caps cdda cddb cdio cdr celt chromaprint cjk clang cli clipboard color-management colord conntrack crypt cups cxx d3d9 dav1d dbus declarative device-mapper dirac djvu dri drm dts dvb dvd dvdr ed25519 editorconfig egl elf emboss encode epub evdev exif faudio fax fbcon ffmpeg fftw filecaps firefox firewalld fish-completion fits flac fontconfig fontforge fortran fribidi gamepad gbm gdal gdbm geoclue geolocation gif git gmp gnupg google googledrive gpg gps graphicsmagick gstreamer gtk gtk3 gzip harfbuzz hdf5 heif http2 ibus iconv icu idn imlib inotify introspection ipv6 jemalloc jpeg jpeg2k json kde kipi kms kwallet ladspa latex lcms ldap libatomic libglvnd libidn2 libinput libnotify libproxy libsecret libsoxr libtirpc libvirt lm-sensors lrz lv2 lvm lz4 lzma lzo mad mariadb markdown mbim mercurial mjpeg mng mobi modemmanager modplug mp3 mp4 mpeg mplayer mpris mtp multilib mysql ncurses netlink networkmanager nls nptl office ofx ogg openal opencl opencv openexr opengl openh264 openmax openmp opus pam pango pcap pch pcre pcre2 pdf pgo phonon pixman pkcs7 plasma pm-utils png policykit postscript ppds prison pulseaudio python qml qrcode qt5 raw rdp readline redfish samba sasl scanner schroedinger screencast sctp sdl sdl2 seccomp semantic-desktop share smartcard snappy sparse speech speex spell ssl startup-notification steamruntime stemmer svg systemd systemtap tbb tcpd teamd telepathy tga theora threads thunderbolt tiff timezone tmux truetype tslib udev udisks uinput unicode unwind upnp upnp-av upower usb utempter v4l v4l2 vaapi vdpau vkd3d vorbis vpx vulkan wasm wavpack wayland webchannel webengine webkit webp widgets wmf wps x264 x265 xattr xcb xcomposite xinerama xkb xml xmp xrandr xscreensaver xv xvid xwayland xz yaml zeroconf zeromq zimg zlib zstd" ABI_X86="64" ADA_TARGET="gnat_2018" ALSA_CARDS="hda-intel" APACHE2_MODULES="authn_core authz_core socache_shmcb unixd actions alias auth_basic authn_alias authn_anon authn_dbm authn_default authn_file authz_dbm authz_default authz_groupfile authz_host authz_owner authz_user autoindex cache cgi cgid dav dav_fs dav_lock deflate dir disk_cache env expires ext_filter file_cache filter headers include info log_config logio mem_cache mime mime_magic negotiation rewrite setenvif speling status unique_id userdir usertrack vhost_alias" CALLIGRA_FEATURES="karbon sheets words" CAMERAS="ptp2" COLLECTD_PLUGINS="df interface irq load memory rrdtool swap syslog" CPU_FLAGS_X86="aes avx avx2 f16c fma3 mmx mmxext pclmul popcnt sha sse sse2 sse3 sse4_1 sse4_2 sse4a ssse3" ELIBC="glibc" ENLIGHTENMENT_MODULES="*" GPSD_PROTOCOLS="ashtech aivdm earthmate evermore fv18 garmin garmintxt gpsclock greis isync itrax mtk3301 nmea ntrip navcom oceanserver oldstyle oncore rtcm104v2 rtcm104v3 sirf skytraq superstar2 timing tsip tripmate tnt ublox ubx" GRUB_PLATFORMS="efi-64" INPUT_DEVICES="joystick libinput" KERNEL="linux" L10N="de de-DE en en-GB ar fa tr ja ko zh zh-CN zh-TW" LCD_DEVICES="bayrad cfontz cfontz633 glk hd44780 lb216 lcdm001 mtxorb ncurses text" LIBREOFFICE_EXTENSIONS="nlpsolver scripting-javascript wiki-publisher" LIRC_DEVICES="devinput" LLVM_TARGETS="AMDGPU BPF RISCV WebAssembly" LUA_TARGET="lua5-2" OFFICE_IMPLEMENTATION="libreoffice" PHP_TARGETS="php7-2" POSTGRES_TARGETS="postgres10 postgres11" PYTHON_SINGLE_TARGET="python3_6" PYTHON_TARGETS="python2_7 python3_6 pypy pypy3" QEMU_SOFTMMU_TARGETS="riscv32 riscv64 x86_64" QEMU_USER_TARGETS="riscv32 riscv64" RUBY_TARGETS="ruby24 ruby25" USERLAND="GNU" VIDEO_CARDS="amdgpu virgl" XTABLES_ADDONS="quota2 psd pknock lscan length2 ipv4options ipset ipp2p iface geoip fuzzy condition tee tarpit sysrq steal rawnat logmark ipmark dhcpmac delude chaos account"
Unset:  CC, CPPFLAGS, CTARGET, CXX, INSTALL_MASK, LC_ALL, LINGUAS, PORTAGE_BINHOST, PORTAGE_BUNZIP2_COMMAND, PORTAGE_COMPRESS, PORTAGE_COMPRESS_FLAGS, PORTAGE_RSYNC_EXTRA_OPTS

=================================================================
                        Package Settings
=================================================================

sys-apps/flatpak-1.6.1::flatpak-overlay was built with the following:
USE="gtk introspection kde (policykit) seccomp -doc -gnome" ABI_X86="(64)"
```

**Comment by devurandom** [comment]
I have been playing a little with `lsns` and `nsenter` from util-linux-2.35.1:
```
❯ sudo lsns | rg steam
4026532987 user       14 376652 dschridde        /usr/libexec/flatpak-bwrap --args 33 steam-wrapper
4026532988 mnt        14 376652 dschridde        /usr/libexec/flatpak-bwrap --args 33 steam-wrapper
4026532989 pid        14 376652 dschridde        /usr/libexec/flatpak-bwrap --args 33 steam-wrapper

❯ nsenter --all --target=376652 /bin/sh
nsenter: reassociate to namespace 'ns/cgroup' failed: Operation not permitted

❯ sudo nsenter --all --target=376652 /bin/sh
nsenter: setgid failed: Invalid argument

❯ sudo nsenter --all --target=376652 --setuid=1000 /bin/sh
nsenter: setgid failed: Invalid argument

❯ sudo nsenter --all --target=376652 --setuid=1000 --setgid=1000 /bin/sh
sh-5.0$ test -d /app && echo inside container || echo broken
inside container
```

**Comment by devurandom** [comment]
More fun with `nsenter`:
```
❯ flatpak ps
Instance   PID    Application             Runtime
3227875980 376644 com.valvesoftware.Steam org.freedesktop.Platform
[...]
❯ sudo nsenter --all --target=376644 --setuid=1000 --setgid=1000 /bin/sh
sh-5.0$ test -d /app && echo inside container || echo broken
broken
```

That would explain why my attempt to enter the namespace in the original post failed: The PID displayed is not PID 1 in the flatpacked application's PID namespace, but somewhere outside of it.

**Comment by alexlarsson** [comment]
Does gentoo use a setuid bwrap or a regular one (i.e. user namespaces)? There are issues with flatpak enter and the setuid case.

**Comment by devurandom** [comment]
No setuid bit is set:
```
❯ ll /usr/libexec/flatpak-bwrap 
Permissions Size User Date Modified Name
.rwxr-xr-x   56k root 15 Feb 21:31  /usr/libexec/flatpak-bwrap
```

Gentoo uses `--without-system-bubblewrap` to make flatpak build its own bubblewrap:
https://github.com/fosero/flatpak-overlay/blob/master/sys-apps/flatpak/flatpak-1.6.2.ebuild#L63-L81

**Comment by refi64** [comment]
#3646 is now fixed in 1.8.0, is this still reproducible for you on that version?

**Comment by devurandom** [comment]
```
❯ flatpak --version
Flatpak 1.8.0
❯ flatpak enter com.valvesoftware.Steam /bin/sh
[📦 com.valvesoftware.Steam ~]$ cat /etc/os-release 
NAME=Freedesktop.org
VERSION="19.08.11 (Flatpak runtime)"
VERSION_ID=19.08.11
ID=org.freedesktop.platform
PRETTY_NAME="Freedesktop.org 19.08.11 (Flatpak runtime)"
BUG_REPORT_URL=https://gitlab.com/freedesktop-sdk/freedesktop-sdk/issues
[📦 com.valvesoftware.Steam ~]$ exit
```
